src/util/small_shared_two_way_ptr.h

```diff
@@ -0,0 +1,287 @@
+/// keep a separate `type` member or using `typeid` or `dynamic_cast`. Moreover,
+/// since the shared pointer is aware of the concrete type of the object being
+/// stored, it can delete it without requiring a virtual destructor or custom
+/// delete function (like std::shared_ptr).
+template <typename U, typename V>
+class small_shared_two_way_ptrt final
+{
+public:
+  typedef decltype(std::declval<U>().use_count()) use_countt;
+
+  typedef small_shared_two_way_pointeet<use_countt> pointeet;
+
+  static_assert(std::is_base_of<pointeet, U>::value, "");
+  static_assert(std::is_base_of<pointeet, V>::value, "");
+
+  small_shared_two_way_ptrt() = default;
+
+  explicit small_shared_two_way_ptrt(U *u) : p(u)
+  {
+    PRECONDITION(u != nullptr);
+    PRECONDITION(u->use_count() == 0);
+
+    p->set_derived1();
+    p->increment_use_count();
+  }
+
+  explicit small_shared_two_way_ptrt(V *v) : p(v)
+  {
+    PRECONDITION(v != nullptr);
+    PRECONDITION(v->use_count() == 0);
+
+    p->set_derived2();
+    p->increment_use_count();
+  }
+
+  small_shared_two_way_ptrt(const small_shared_two_way_ptrt &rhs) : p(rhs.p)
+  {
+    PRECONDITION(is_same(rhs));
+
+    if(p)
+    {
+      p->increment_use_count();
+    }
+  }
+
+  small_shared_two_way_ptrt(small_shared_two_way_ptrt &&rhs)
+  {
+    PRECONDITION(is_same(rhs));
+
+    swap(rhs);
+  }
+
+  small_shared_two_way_ptrt &operator=(const small_shared_two_way_ptrt &rhs)
+  {
+    PRECONDITION(is_same(rhs));
+
+    small_shared_two_way_ptrt copy(rhs);
+    swap(copy);
+    return *this;
+  }
+
+  small_shared_two_way_ptrt &operator=(small_shared_two_way_ptrt &&rhs)
+  {
+    PRECONDITION(is_same(rhs));
+
+    swap(rhs);
+    return *this;
+  }
+
+  ~small_shared_two_way_ptrt()
+  {
+    if(!p)
+    {
+      return;
+    }
+
+    auto use_count = p->use_count();
+
+    if(use_count == 1)
+    {
+      if(p->is_derived1())
+      {
+        U *u = static_cast<U *>(p);
+        delete u;
+      }
+      else
+      {
+        V *v = static_cast<V *>(p);
+        delete v;
+      }
+    }
+    else
+    {
+      p->decrement_use_count();
+    }
+  }
+
+  void swap(small_shared_two_way_ptrt &rhs)
+  {
+    PRECONDITION(is_same(rhs));
+
+    std::swap(p, rhs.p);
+  }
+
+  use_countt use_count() const
+  {
+    return p ? p->use_count() : 0;
+  }
+
+  bool is_derived1() const
+  {
+    return p == nullptr || p->is_derived1();
+  }
+
+  bool is_derived2() const
+  {
+    return p == nullptr || p->is_derived2();
+  }
+
+  pointeet *get() const
+  {
+    return p;
+  }
+
+  U *get_derived1() const
+  {
+    PRECONDITION(is_derived1());
+
+    return static_cast<U *>(p);
+  }
+
+  V *get_derived2() const
+  {
+    PRECONDITION(is_derived2());
+
+    return static_cast<V *>(p);
+  }
+
+  bool is_same(const small_shared_two_way_ptrt &other) const
+  {
+    const bool b1 = p == nullptr;
+    const bool b2 = other.p == nullptr;
+
+    if(b1 || b2)
+      return true;
```

Comment:
How does this work? Why if either of these is true, then are they the same? Isn't it the case that the object is in a compromised logical state if `p == nullptr` or that the function is semantically void if `other.p == nullptr`?

Reply:
The function checks whether both pointers are convertible to either `U *` or `V *`, so since `nullptr` can be converted to both we return true here. Maybe I should change the name of the function to `is_same_type()` or something like that. Any suggestions?

Reply:
Yes, `is_same_type` sounds better to me.